### DISCUSSION TITLE
Charts: Custom Calculated Percentage Metrics not behaving as expected in Tables Chart

### DISCUSSION THREAD

**Discussion by jacobcroope** [apache/superset#18843] [Q&A / Help] [answered]
We are interested in calculating a fair share metric which requires dividing two percentages within a table. The percentage metrics correctly calculate a % of what is displayed in a table dynamically with filters which is critical to our application. Here is a screenshot of the type of visualization I am trying to make. I would expect the value to be an index around 1. For Wii Sports with 20.38% of North American Market and 22.39% of Global Market the value should be 91%

![image](https://user-images.githubusercontent.com/1770898/154991299-0c189ced-9c83-4f38-9bc1-5306bb1b4072.png)

The SQL statement is dividing them before calculating percentages.
![image](https://user-images.githubusercontent.com/1770898/154992363-3875c941-da87-41f9-a621-38c7de62732d.png)

In PowerBi, Tableu, and Excel there is a way to accomplish this. Is there a way to do this in Superset I am unaware of or is this something that needs to be added to the visualization?

Other operations don't provide the expected results either. 
![image](https://user-images.githubusercontent.com/1770898/154993162-606a1567-cd41-4470-be64-4d35628d85f7.png)

### Environment
Superset - Current as of today. 
### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [X] I have reproduced the issue with at least the latest released version of superset.
- [X] I have checked the issue tracker for the same issue and I haven't found one similar.

**Comment by villebro** [comment]
If you want to calculate the proportion of one aggregate over another one, I believe you should use `SUM` rather than `AVG`? In other words, to calculate the total amount of sales in Nort America as a proportion of global sales, I would do `SUM(na_sales)/SUM(global_sales)`, not `AVG(na_sales)/AVG(global_sales)`:

![image](https://user-images.githubusercontent.com/33317356/155108302-c089db14-27ad-4452-8061-18714e1fde4f.png)

Or am I misunderstanding something?

**Reply by jacobcroope** [reply]
I didn't think to check sum because we need averages as we are interested in displaying averages of a variable time grain. Hourly, Daily, Weekly, Monthly etc. However it doesn't seem to change the behavior. 
![image](https://user-images.githubusercontent.com/1770898/155165939-d4c1713f-bf6f-495a-b39b-3ff63c9a558b.png)


Coming from the excel/PowerBI world you would create your measure first which you can then perform calculations/manipulations on. Taking that data set into excel I can create what I want with data model pivot table functions. I also tried the superset PivotV2 version to see if it had a better way. It does do table subtotals which is useful but it does not implement what I would consider the "standard" behavior as defined by excel. 

![Excel_Fair_Share](https://user-images.githubusercontent.com/1770898/155164306-058dd42b-f8b9-4b9b-9bcd-b3d3ffbde9a2.png)
![Excel_Fair_Share_2](https://user-images.githubusercontent.com/1770898/155164335-49bd337a-46c5-4120-a60b-2857345972c5.png)

This seems like it should be moved back to an issue/feature request as it would be the expected behavior by most users. I also definitely need the ability to do this so guidance on where the percentage metrics code is located would be helpful in giving me a starting point.

**Comment by villebro** [comment, marked as answer]
Converting to discussion

**Reply by villebro** [reply]
It's good to remember, that any metric you can express with a SQL expression, you can do in Superset. If you want to calculate the ratio of the ratios of `na_sales` and `global_sales` (assuming I understood your intent correctly), you probably want to use window functions. Breaking it down:

`single_game_na_sales = SUM(na_sales)`
`total_na_sales = SUM(SUM(na_sales)) OVER ()`
`single_game_global_sales = SUM(global_sales)`
`total_global_sales = SUM(SUM(global_sales)) OVER ()`

then the ratios:

`na_ratio = single_game_na_sales / total_na_sales`
`global_ratio = single_game_global_sales / total_global_sales`

and finally:

`fair_share = na_ratio / global_ratio` i.e.
`fair_share = (SUM(na_sales) / SUM(SUM(na_sales)) OVER ()) / (SUM(global_sales) / SUM(SUM(global_sales)) OVER ())`

Here's an example chart with all those metrics:

<img width="1493" alt="image" src="https://user-images.githubusercontent.com/33317356/155330006-4ea4dd05-efc5-4a68-8d34-338ccd2ea638.png">

If you want to only calculate the ratios on the top N rows then you need to do some additional petritioning or use a subquery (or even Jinja templating if you need to get even more advanced). But again, if you can do it with SQL, you can do it with Superset.

> I didn't think to check sum because we need averages as we are interested in displaying averages of a variable time grain. Hourly, Daily, Weekly, Monthly etc. However it doesn't seem to change the behavior.

P.S. I didn't see any use of time grains here, or did I miss something?

**Reply by jacobcroope** [reply]
@villebro Appreciate the detailed response and the SQL guidance is very helpful for next steps. I see that your provided solution does handle filtering and correctly calculates what I am looking for. 

It seems like perhaps we are getting off track here however. 

I am only using this as an example because the data set I am utilizing internally is under a lot of NDA's. I'll look through the examples data set for something that has a time component and groupings as a better example to illustrate my point.  

For clients/end users who live 99% in excel and are not going to learn SQL at this level it seems very important for "Percentage Metrics" which are natively filter aware to behave in a manner they are most used to. 

`(SUM(na_sales) / SUM(SUM(na_sales)) OVER ()) / (SUM(global_sales) / SUM(SUM(global_sales)) OVER ())`

Is a valid solution for this particular issue but I think the better solution would be to make it so Percentage Metrics work with. 

`(SUM(na_sales)/SUM(global_sales))  `

For most of my internal and external users this is as complicated of a query as they are going to be interested in writing. I am a full time analyst and was not able to solve this at first because I have been trained by the predominant ecosystem to interact with these tools in this manner. 

Other features like subtotaling on Groups is on my feature list of things that need to be implemented for tables to work for us.

Is there a larger discussion on how these features should be implemented? I would not want to contribute breaking changes to the behavior of percentage metrics/add subtotals to tables and have to maintain a separate branch if it makes more sense to just make a new table visualization that performs these operations in an excel like manner.

**Reply by ramrajsingh** [reply]
(https://github.com/apache/superset/issues/25747#issuecomment-2446315209)
